src/planning/src/planning/velocity_planning.cpp

```diff
@@ -0,0 +1,87 @@
+#include "planning/velocity_planning.hpp"
+
+double VelocityPlanning::find_circle_center(PathPoint &point1, PathPoint &point2,
+                                            PathPoint &point3) {
+  double x1 = point1.position.x;
+  double y1 = point1.position.y;
+  double x2 = point2.position.x;
+  double y2 = point2.position.y;
+  double x3 = point3.position.x;
+  double y3 = point3.position.y;
+
+  PathPoint mid1 = PathPoint((x1 + x2) / 2, (y1 + y2) / 2, 0);
+  PathPoint mid2 = PathPoint((x2 + x3) / 2, (y2 + y3) / 2, 0);
+  double slope1 = (x2 != x1) ? ((y2 - y1) / (x2 - x1)) : MAXFLOAT;
+  double slope2 = (x3 != x2) ? ((y3 - y2) / (x3 - x2)) : MAXFLOAT;
+  double slope1_perpendicular = -1 / slope1;
+  double slope2_perpendicular = -1 / slope2;
+  double center_x = (slope1_perpendicular * mid1.position.x -
+                     slope2_perpendicular * mid2.position.x + mid2.position.y - mid1.position.y) /
+                    (slope1_perpendicular - slope2_perpendicular);
+  double center_y = slope1_perpendicular * (center_x - mid1.position.x) + mid1.position.y;
+  double radius = sqrt(pow(center_x - x2, 2) + pow(center_y - y2, 2));
+  return radius;
+}
+
+void VelocityPlanning::point_speed(std::vector<double> &radiuses, std::vector<double> &velocities) {
+  for (int i = 0; i < static_cast<int>(radiuses.size()); i++) {
+    double velocity = sqrt(abs(config_.braking_acceleration_ * radiuses[i]));
+    velocities.push_back(std::max(velocity, config_.safety_speed_));
+  }
+  velocities.back() = config_.safety_speed_;
+
+  return;
+}
+
+void VelocityPlanning::speed_limiter(std::vector<PathPoint> &points,
+                                     std::vector<double> &velocities) {
+  for (int i = static_cast<int>(points.size()) - 2; i >= 0; i--) {
+    double distance = 0;
+    double max_speed = velocities[i];
+
+    for (int j = i + 1; j < static_cast<int>(points.size())-1; j++) {
+      // Calculate segment distance
```

Comment:
I don't think we need to do this inner loop. Imagine we have an ordered path going from points P[0] to P[n]. 
We start by checking in the distance between P[n-1] and P[n] we can go from the maximum speed at P[n-1] to the maximum speed at P[n]. After that, when we check if we can get from the maximum speed at P[n-2] in the space between P[n-2] and P[n-1], we don't need to check the same thing for the speed at P[n] because we already know that from P[n-1] to P[n] there's no conflict between the maximum speeds. 

In short, I think we can do this in O(n)

Reply:
Done